shared-libs/contacts/src/places.js

```diff
@@ -104,15 +106,25 @@ const createPlace = async (place) => {
     place.parent = lineage.minifyLineage(place.parent);
   }
 
-  const response = await db.medic.post(place);
   if (!contact) {
-    return response;
+    return await db.medic.post(place);
+  }
+
+  if (_.isString(contact)) {
+    const person = await people.getOrCreatePerson(contact);
+    place.contact = person._id;
+    return await db.medic.post(place);
+  }
+
+  if (!_.isObject(contact)) {
+    return Promise.reject({ code: 400 });
   }
-  const placeUUID = response.id;
 
-  contact.place = placeUUID;
+  contact = await preparePlaceContact(contact);
+  const placeResponse = await db.medic.post(place);
+  contact.place = placeResponse.id;
   const person = await people.getOrCreatePerson(contact);
-  const result = await updatePlace(placeUUID, { contact: person._id });
+  const result = await updatePlace(placeResponse.id, { contact: person._id });
```

Comment:
Just suggesting a slightly more compact way of writing this code: 
```

const prepareContact = async (contact) => {
  if (!contact) {
    return;
  }

  if (_.isString(contact)) {
    return await people.getOrCreatePerson(contact);
  }

  if (_.isObject(contact)) {
    contact.type = contact.type || people._getDefaultPersonType();
    const errStr = people._validatePerson(contact);
    if (errStr) {
      return err(errStr);
    }
    return contact;
  }
}

const createPlace = async (place) => {
  await module.exports._validatePlace(place);
  const contact = await prepareContact(place.contact);
  delete place.contact;

  const date = place.reported_date ? utils.parseDate(place.reported_date) : new Date();
  place.reported_date = date.valueOf();
  if (place.parent) {
    place.parent = lineage.minifyLineage(place.parent);
  }

  if (!contact) {
    return await db.medic.post(place);
  }

  if (contact._id) {
    place.contact = contact._id;
    return await db.medic.post(place);
  }

  const placeResponse = await db.medic.post(place);
  contact.place = placeResponse.id;
  const person = await people.getOrCreatePerson(contact);
  const result = await updatePlace(placeResponse.id, { contact: person._id });
  return { ...result, contact: { id: person._id } };
};
```

What do you think?

Reply:
Looks good. I was initially trying to keep `prepareContact` as an async wrapper around `validatePerson` that throws on err ~and to please sonar~. One concern with this suggestion is if a user included an `_id` in the contact object then i think we'd assign a non-existent contact to the place

Reply:
> One concern with this suggestion is if a user included an _id in the contact object then i think we'd assign a non-existent contact to the place

I'm aware. I'm sorry I didn't mention this in a comment or something, but that check was a placeholder for: assess whether this is a new or existing contact. I'm hoping there is some clean way to do that.

Reply:
Cool cool, i decided to return an object instead with an `exists` field so we don't check any payload fields

Reply:
Cool! Thanks!